### ISSUE TITLE
Deposit contract log processing

### ISSUE THREAD

**Issue by paulhauner** [sigp/lighthouse#102]
## Description

This is a "mega issue" which contains multiple discreet tasks. Mega issues exist to avoid cluttering the issues page a multitude of small tasks.

If you wish to work on one of these tasks, comment below and a maintainer will break the task out into a separate issue.

## Log Processing

Validators enter the 2.0 chain through "deposit" smart contract on the 1.0 chain. Therefore, it is necessary for Lighthouse to connect to an existing 1.0 client (e.g., geth, parity) and process logs from the 1.0 chain.

There are two obvious methods for obtaining the logs:

1. Poll the 1.0 client and store the logs in disk/memory.
2. Request the logs whenever they are required.

Option (2) seems to be the most attractive at this point. it involves less moving parts (no polling logic, data storage, etc.) However we might run into issues if there's significant lag in hitting the client, but I can't see this being the case

Assuming log-obtaining-method (2), we need the following:

- An ability to determine the range of blocks from which we should withdraw logs. (I.e., "get logs from block `x` to `y`").
- A method to parse a `Deposit` log into a `ValidatorRegistration`.
- The ability to process a `ChainStart` log.

Furthermore, we will need an integration testing environment to check integration between Lighthouse and the 1.0 client (e.g., Geth, Parity-Ethereum, etc.)

## Immediate Tasks

- Consider the two log-obtaining methods and come to an agreement upon this.

**Comment by ralexstokes** [comment]
i'm w/ you in leanings towards not storing a bunch of logs and handling all the serde and related disk IO... my concern w/ "request on-demand" just comes down to the event where there are many Deposit messages in a given epoch and then Lighthouse slows down waiting on an Eth1.0 endpoint. we would need to benchmark this to know if its worth the work.

another option which afaik is possible w/ any of the Eth1.0 clients which is to set up a notification interface for the relevant `Deposit` events. with this option we just keep an in-memory cache of the most recent `Deposit` events (up to some number or size in memory) and can make one heavy call (enough to fill the cache) upon boot, then w/ some cache policy fill events as they come in via subscription.

**Comment by paulhauner** [comment]
I agree, caching will probably be ideal.

We could take an approach where we implement without caching at first, but we design the interface such that caching can be implemented later with minimal changes required to calling functions.
